chained/shadowsocks_impl.go

```diff
@@ -75,11 +81,49 @@ func newShadowsocksImpl(name, addr string, pc *config.ProxyConfig, reportDialCor
 	source := mrand.NewSource(seed)
 	rng := mrand.New(source)
 
+	tlsConfig := &tls.Config{}
+	if cert := pc.Cert; cert != "" {
+		block, rest := pem.Decode([]byte(cert))
+		if block == nil {
+			return nil, errors.New("failed to decode proxy certificate as PEM block")
+		}
+
+		if len(rest) > 0 {
+			return nil, errors.New("unexpected extra data in proxy certificate PEM")
+		}
+
+		if block.Type != "CERTIFICATE" {
+			return nil, errors.New("expected certificate in PEM block")
+		}
+		proxyCert := block.Bytes
+
+		verifyPeerCert := func(peerCerts [][]byte, _ [][]*x509.Certificate) error {
+			if len(peerCerts) == 0 {
+				return errors.New("no peer certificate")
+			}
+			if !bytes.Equal(peerCerts[0], proxyCert) {
+				return errors.New("peer certificate does not match expected")
+			}
+			return nil
+		}
+
+		certPool := x509.NewCertPool()
+		certPool.AppendCertsFromPEM([]byte(cert))
+		ip, _, _ := net.SplitHostPort(addr)
+		tlsConfig = &tls.Config{
+			RootCAs:               certPool,
+			ServerName:            ip,
+			InsecureSkipVerify:    true,
+			VerifyPeerCertificate: verifyPeerCert,
```

Comment:
So... I might have led you astray.  Thinking about this more, I think we might be able to just include the proxy certificate in `RootCAs`, as you have done, and skip the cert pinning / `VerifyPeerCertificate` stuff.  The proxy certificate is self-signed, so I think verification should pass if the proxy certificate is in the CA pool.

The reason we cannot do this normally is because we want to specify a bogus SNI (to throw off censors).  This bogus SNI causes normal verification to fail.  In this case, the TLS layer is wrapped in Shadowsocks, so the SNI will not be visible anyway.

Sorry for the mix-up!

Reply:
In short: I think we can do away with `VerifyPeerCertificate` and `InsecureSkipVerify` and keep everything else as-is.